hibernate-core/src/test/java/org/hibernate/jpa/test/enhancement/TestDirtyCheckOnLazy.java

```diff
@@ -0,0 +1,174 @@
+import java.util.Arrays;
+import java.util.Map;
+
+import static org.hibernate.testing.transaction.TransactionUtil.doInJPA;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@TestForIssue( jiraKey = "HHH-14244" )
+@RunWith( BytecodeEnhancerRunner.class )
+public class TestDirtyCheckOnLazy extends BaseEntityManagerFunctionalTestCase {
+
+    private EntityWithLazyProperty entity;
+
+    @Override
+    public Class[] getAnnotatedClasses() {
+        return new Class[]{EntityWithLazyProperty.class};
+    }
+
+    @Override
+    protected void addConfigOptions(Map options) {
+        options.put( AvailableSettings.CLASSLOADERS, getClass().getClassLoader() );
+    }
+
+    @Before
+    public void prepare() throws Exception {
+        EntityPersister ep = entityManagerFactory().getMetamodel().entityPersister( EntityWithLazyProperty.class.getName() );
+        assertTrue( ep.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading() );
+
+        byte[] testArray = new byte[]{0x2A};
+
+        doInJPA( this::entityManagerFactory, em -> {
+            //persist the test entity.d
+            entity = new EntityWithLazyProperty();
+            entity.setSomeField( "TEST" );
+            entity.setLazyData( testArray );
+            em.persist( entity );
+        } );
+
+        checkLazyField( entity, testArray );
+    }
+
+    /**
+     * Check the dirty check is actually taking place, preventing update of the lazy field
+     */
+    @Test
+    public void testNoUpdate() {
+        byte[] testArray = new byte[]{0x2A};
+
+        doInJPA( this::entityManagerFactory, new JPATransactionVoidFunction() {
+            @Override
+            public void accept(EntityManager em) {
+                entity = em.find( EntityWithLazyProperty.class, entity.id );
+                assertFalse( Hibernate.isPropertyInitialized( entity, "lazyData" ) );
+                entity.setSomeField( "TEST1" );
+            }
+
+            @Override
+            public void afterTransactionCompletion() {
+                assertFalse( Hibernate.isPropertyInitialized( entity, "lazyData" ) );
+            }
+        } );
+    }
+
+    /**
+     * Update a lazy field to a value and check the value is read back.
+     */
+    @Test
+    public void testUpdate() {
+        byte[] testArray = new byte[]{0x44};
+
+        doInJPA( this::entityManagerFactory, new JPATransactionVoidFunction() {
+            @Override
+            public void accept(EntityManager em) {
+                entity = em.find( EntityWithLazyProperty.class, entity.id );
+                assertFalse( Hibernate.isPropertyInitialized( entity, "lazyData" ) );
+                entity.setLazyData(testArray);
+            }
+
+            @Override
+            public void afterTransactionCompletion() {
+                assertTrue( Hibernate.isPropertyInitialized( entity, "lazyData" ) );
+            }
+        } );
+
+        checkLazyField( entity, testArray );
+    }
```

Comment:
seems the above two methods were copied from `TestLazyPropertyOnPreUpdate`. Are they relevant to the nullness issue? We might delete them if not.